### PR TITLE
GCM topic support and fallbacks

### PR DESCRIPTION
**Pull request by matzew**
Done for [AGPUSH-1454](https://issues.jboss.org/browse/AGPUSH-1454)

@secondsun @matskiv this is the PR for the 1.1.x series. Once this is merged, I will get https://github.com/aerogear/aerogear-unifiedpush-server/pull/708 updated and land it on master

### CODE REVIEW

File: push/model/src/main/java/org/jboss/aerogear/unifiedpush/message/Criteria.java [hunk 1]
```diff
@@ -82,6 +83,37 @@ public void setVariants(List<String> variants) {
         this.variants = variants;
     }
 
+    /**
+     * Helper method to check if only categories are applied. Usful in GCM land, where we use topics
+     */
+    @JsonIgnore
+    public boolean isCategoryOnly() {
```
Comment (matzew) [inline push/model/src/main/java/org/jboss/aerogear/unifiedpush/message/Criteria.java]: Perhaps these methods should be moved to a util, next to TokenLoader?

Not sure if this is generally helpful on the `Criteria` API class

Comment (matzew) [inline push/model/src/main/java/org/jboss/aerogear/unifiedpush/message/Criteria.java]: @secondsun I thought about this, and will move this into a (GCM Topic) Util class.

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java [hunk 2]
```diff
@@ -167,4 +194,20 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
             }
         }
     }
-}
\ No newline at end of file
+
+    private Set<String> extractTopicsFrom(final Criteria criteria, final String variantID) {
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: @secondsun this function is another candidate for the (GCM Topic) Util class. 

I will update the PR...

**Comment by matzew** [comment]
@secondsun ok, moved the TokenLoader around and was adding a util, instead of blowing up the public Criteria API (used on the java sender too)

**Comment by matzew** [comment]
Some testing info, @secondsun @matskiv

I was using an older app (e.g. old [hello world](https://github.com/jboss-mobile/unified-push-helloworld/tree/master/android), which uses GCM register to get a legacy token)

1)  sending a push, with a category:
```
17:25:00,293 INFO  [org.jboss.aerogear.unifiedpush.message.sender.GCMPushNotificationSender] (Thread-46 (HornetQ-client-global-threads-1851833922)) Sent push notification to GCM topic: &#x2F;topics&#x2F;soccer
...
17:25:00,320 INFO  [org.jboss.aerogear.unifiedpush.message.sender.GCMPushNotificationSender] (Thread-24 (HornetQ-client-global-threads-1851833922)) Sent push notification to GCM Server for 1 registrationIDs
```

It first sends to the topic, but also looks for legacy tokens, as fallback. Message is delivered (since my device is in the 'soccer' category)

✅ 

2) Similar, with with no criteria:
```
17:33:56,432 INFO  [org.jboss.aerogear.unifiedpush.message.sender.GCMPushNotificationSender] (Thread-46 (HornetQ-client-global-threads-1851833922)) Sent push notification to GCM topic: &#x2F;topics&#x2F;50fdef1f-e9f6-435a-911d-5cc111e71a5f
...
17:33:56,448 INFO  [org.jboss.aerogear.unifiedpush.message.sender.GCMPushNotificationSender] (Thread-37 (HornetQ-client-global-threads-1851833922)) Sent push notification to GCM Server for 1 registrationIDs
```

It first sends to the 'global; topic, but also looks for legacy tokens, as fallback. Message is delivered.

✅ 

3) Now, trying a category which my device is not in:

```
17:37:08,229 INFO  [org.jboss.aerogear.unifiedpush.message.sender.GCMPushNotificationSender] (Thread-25 (HornetQ-client-global-threads-1851833922)) Sent push notification to GCM topic: &#x2F;topics&#x2F;dasdsadsadsa
17:37:08,231 WARNING [org.jboss.aerogear.unifiedpush.message.TokenLoader] (Thread-58 (HornetQ-client-global-threads-1851833922)) Check your push query: Not a single token was loaded from the DB!
```

First sends to the given topic (or category), afterwards looks for the tokens - but there is no token, as no device is registered for this category. Ergo, no message is delivered 


✅

**Comment by secondsun** [comment]
LGTM +1

**Comment by matzew** [comment]
feel free to merge!

On Tuesday, 3 May 2016, Hoyt Summers Pittman <notifications@github.com>
wrote:

> LGTM +1
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/709#issuecomment-216627833>
>


-- 
Sent from Gmail Mobile

**Comment by matzew** [comment]
landed - will send different PR for `master`

**Pull request merged**